plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "got %lu bytes, less than GAIA header",
+			    *read_len);
+		return FALSE;
+	}
+	memcpy(data_in,
+	       buf + FU_QC_S5GEN2_GAIA_V3_HDR_SZ,
+	       (*read_len) - FU_QC_S5GEN2_GAIA_V3_HDR_SZ);
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_req_connect(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_upgrade_connect_cmd_new();
+	g_autoptr(GByteArray) validate = NULL;
+
+	fu_struct_qc_gaia_v3_upgrade_connect_cmd_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_notify_acquire(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_upgrade_connect_ack_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_req_disconnect(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_upgrade_disconnect_cmd_new();
+	g_autoptr(GByteArray) validate = NULL;
+
+	fu_struct_qc_gaia_v3_upgrade_disconnect_cmd_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_upgrade_disconnect_ack_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_release(self, error);
+}
+
+static gsize
+fu_qc_s5gen2_ble_device_data_size(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	/* TODO: atm for GAIA only */
+	return self->mtu - FU_STRUCT_QC_DATA_SIZE - FU_QC_S5GEN2_GAIA_V3_HDR_SZ - 3;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_get_api(FuQcS5gen2BleDevice *self, GError **error)
+{
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_api_req_new();
+	g_autoptr(GByteArray) resp = NULL;
+
+	fu_struct_qc_gaia_v3_api_req_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	resp = fu_struct_qc_gaia_v3_api_parse(buf, read_len, 0, error);
+	if (resp == NULL)
+		return FALSE;
+
+	self->api_version.major = fu_struct_qc_gaia_v3_api_get_major(resp);
+	self->api_version.minor = fu_struct_qc_gaia_v3_api_get_minor(resp);
```

Comment:
i don't understand why we're saving the api version in `self` -- it's not used anywhere other than to_string. I think in that case we could just set is as fu_device_set_bootloader_version{_raw?) or something like that?

Reply:
hm... Correct. 
It is needed to validate if the device support commands. moved to local variables.